### PR TITLE
GEOMESA-1635 Java and Python GeoMesaSpark bindings

### PR DESCRIPTION
**Pull request by tkunicki**
- Additional Accumulo InputFormats
  - String (GeoJSON)
  - Array[Byte] (Kryo Serialized Simple Feature)
  - Map (Scala)
- Type Parameterized `SpatialRDDProvider.rdd(...)`
- implement Java wrapper

### CODE REVIEW

File: geomesa-spark/geomesa-spark-converter/src/main/scala/org/locationtech/geomesa/spark/converter/ConverterSpatialRDDProvider.scala [hunk 1]
```diff
@@ -43,14 +44,15 @@ import scala.util.control.NonFatal
...
 
-  override def rdd(conf: Configuration,
-                   sc: SparkContext,
-                   params: Map[String, String],
-                   query: Query): RDD[SimpleFeature] = {
+  override def rdd[T : ClassTag](conf: Configuration,
+                                 sc: SparkContext,
+                                 params: Map[String, String],
+                                 query: Query)(implicit default: T := SimpleFeature): RDD[T] = {
     val sftName = params.get(FeatureNameKey).orElse(Option(query.getTypeName)).orNull
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-converter/src/main/scala/org/locationtech/geomesa/spark/converter/ConverterSpatialRDDProvider.scala]: we should validate that the classtag is `SimpleFeature` up front - anything else will blow up down in the `.asInstanceOf` i believe

File: geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSpark.scala [hunk 2]
```diff
@@ -65,3 +68,16 @@ object CaseInsensitiveMapFix {
     case _ => new MapWrapper[A,B](m) with MapWrapperFix[A, B]
   }
 }
+
+object TypeDefault {
+
+  class :=[P,D]
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSpark.scala]: could you add some comments? hard to follow what this is doing

File: geomesa-spark/geomesa-spark-geotools/src/main/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProvider.scala [hunk 3]
```diff
@@ -23,47 +23,49 @@ import org.locationtech.geomesa.utils.io.CloseQuietly
...
-  override def rdd(conf: Configuration,
-                   sc: SparkContext,
-                   dsParams: Map[String, String],
-                   query: Query): RDD[SimpleFeature] = {
-    val ds = DataStoreFinder.getDataStore(dsParams)
+  override def rdd[T : ClassTag](conf: Configuration,
+                                  sc: SparkContext,
+                                  params: Map[String, String],
+                                  query: Query)(implicit default: T := SimpleFeature): RDD[T] = {
+    val ds = DataStoreFinder.getDataStore(params)
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-geotools/src/main/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProvider.scala]: should validate the classtag up front

File: geomesa-accumulo/geomesa-accumulo-spark/src/main/scala/org/locationtech/geomesa/spark/accumulo/AccumuloSpatialRDDProvider.scala [hunk 4]
```diff
@@ -26,43 +27,55 @@ import org.locationtech.geomesa.accumulo.index.EmptyPlan
...
-  def rdd(conf: Configuration,
-          sc: SparkContext,
-          dsParams: Map[String, String],
-          query: Query): RDD[SimpleFeature] = {
-    val ds = DataStoreFinder.getDataStore(dsParams).asInstanceOf[AccumuloDataStore]
-    val username = AccumuloDataStoreParams.userParam.lookUp(dsParams).toString
-    val password = new PasswordToken(AccumuloDataStoreParams.passwordParam.lookUp(dsParams).toString.getBytes)
+  def inputFormat[T : ClassTag](): Class[InputFormat[Text, T]] = {
+    (classTag[T] match {
+      case c if c == classTag[SimpleFeature] => classOf[GeoMesaAccumuloInputFormat]
```
Comment (elahrvivaz) [inline geomesa-accumulo/geomesa-accumulo-spark/src/main/scala/org/locationtech/geomesa/spark/accumulo/AccumuloSpatialRDDProvider.scala]: to follow up on my earlier comment - instead of this block, have something like:
```
sealed trait SpatialRDDFormat[T]
case object SimpleFeatureFormat extends SpatialRDDFormat[SimpleFeature]
...
def rdd(conf: Configuration, sc: ..., params: ..., query: ...)(implicit format: SpatialRDDFormat[T] = SimpleFeatureFormat): RDD[T]
```

**Comment by aheyne** [comment]
Looks good just missing a few new lines and license headers in the .java files.

File: geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSpark.scala [hunk 5]
```diff
@@ -8,32 +8,76 @@
...
+
+import scala.collection.JavaConversions._
+import scala.reflect._
 
 trait SpatialRDDProvider {
+  import TypeDefault._
+
+  type GeoJSONString = String
+
+  def toScalaMap(in: SpatialRDD): RDD[Map[String, AnyRef]] = {
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSpark.scala]: we should make this private or protected since looks like they aren't meant to be called publicly

File: geomesa-accumulo/geomesa-accumulo-spark/src/main/scala/org/locationtech/geomesa/spark/accumulo/AccumuloSpatialRDDProvider.scala [hunk 6]
```diff
@@ -26,43 +27,54 @@ import org.locationtech.geomesa.accumulo.index.EmptyPlan
...
 
-  def rdd(conf: Configuration,
-          sc: SparkContext,
-          dsParams: Map[String, String],
-          query: Query): RDD[SimpleFeature] = {
-    val ds = DataStoreFinder.getDataStore(dsParams).asInstanceOf[AccumuloDataStore]
-    val username = AccumuloDataStoreParams.userParam.lookUp(dsParams).toString
-    val password = new PasswordToken(AccumuloDataStoreParams.passwordParam.lookUp(dsParams).toString.getBytes)
+//  def inputFormat[T : ClassTag](): Class[InputFormat[Text, T]] = {
+//    (classTag[T] match {
```
Comment (elahrvivaz) [inline geomesa-accumulo/geomesa-accumulo-spark/src/main/scala/org/locationtech/geomesa/spark/accumulo/AccumuloSpatialRDDProvider.scala]: remove commented out code

File: geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSpark.scala [hunk 7]
```diff
@@ -8,32 +8,76 @@
...
 
-  def rdd(conf: Configuration,
-          sc: SparkContext,
-          dsParams: Map[String, String],
-          query: Query): RDD[SimpleFeature]
+  def rdd[T : ClassTag](conf: Configuration,
+             sc: SparkContext,
+             params: Map[String, String],
+             query: Query)(implicit default: T := SimpleFeature): RDD[T] = {
+      val tf = transformFunc[T]
```
Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSpark.scala]: would it be worthwhile to override this in the accumulo class to use the various input formats directly and save the map step?

**Comment by elahrvivaz** [comment]
i like the formats stuff - looks good

File: geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala [hunk 8]
```diff
@@ -43,7 +44,7 @@ class GeoToolsSpatialRDDProviderTest extends Specification {
       val ds = DataStoreFinder.getDataStore(dsParams)
       ingestChicago(ds)
 
-      val rdd = GeoMesaSpark(dsParams).rdd(new Configuration(), sc, dsParams, new Query("chicago"))
+      val rdd = GeoMesaSpark(dsParams).rdd[SimpleFeature](new Configuration(), sc, dsParams, new Query("chicago"))
```
Comment (jnh5y) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: Is adding the type argument necessary?

Comment (tkunicki) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: in the absence of an implicit val specifying transform then yes.

Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: could we make the `.rdd` method have a default implicit transform to `SimpleFeature`?

Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: maybe just have to add an implicit val in the companion object - not sure how that plays with the implicit defs:
http://stackoverflow.com/questions/12767074/how-to-provide-default-value-for-implicit-parameters-at-class-level

Comment (tkunicki) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: i will try, but this makes me nervous
```
implicit transform: SpatialRDD => RDD[T] = toSimpleFeatureWithKryoRegistration.asInstanceOf[SpatialRDD => RDD[T]])
```

Comment (tkunicki) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: @elahrvivaz we already have a suite of implicits in companion class.  an alternative is to try and play with scope precedence.  e.g. find SimpleFeature first, then find others.

Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: so right now adding the type parameter (`.rdd[SimpleFeature]`) is causing it to find the equivalent implicit def, is that correct?

Comment (tkunicki) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: yes, the issue is that in the absence of typing that more than one implicit can be applied given the current precedence.

Comment (elahrvivaz) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: what about overloading the method? have a `.rdd` with no implicit argument that calls `.rdd[T]` with the simple feature implicit value. (I'm not sure if that would actually work)

Comment (tkunicki) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: it won't, this is the reason the previous method was renamed from `rdd` to `read`

File: geomesa-spark/geomesa-spark-core/src/main/resources/geomesa/spark.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
+from py4j.java_gateway import java_import
```
Comment (clgy15) [inline geomesa-spark/geomesa-spark-core/src/main/resources/geomesa/spark.py]: Have we thought about producing a pip package? Refactor the python stuff into a new module, that wraps this up as a pip project. We can include the accumulo-runtime jar in the pip (I think), and so we can distribute everything except for the SPARK dist in one easy to use package.

Comment (clgy15) [inline geomesa-spark/geomesa-spark-core/src/main/resources/geomesa/spark.py]: Also makes this very python-esqe

File: geomesa-hbase/geomesa-hbase-tools/src/main/resources/org/locationtech/geomesa/hbase/tools/ingest-libjars.list [hunk 10]
```diff
@@ -7,6 +7,10 @@ geomesa
 gson
 gt
 hbase
+hbase-client
+hbase-server
+hbase-protocol
+htrace
```
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-tools/src/main/resources/org/locationtech/geomesa/hbase/tools/ingest-libjars.list]: Hmm... are these in this PR on purpose?  (I'm wondering a few things got merged together.)

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/IndexAdapter.scala [hunk 11]
```diff
@@ -13,7 +13,7 @@
 
 package org.locationtech.geomesa.index.index
 
-import com.typesafe.scalalogging.{LazyLogging, Logger}
+import com.typesafe.scalalogging.LazyLogging
```
Comment (jnh5y) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/IndexAdapter.scala]: ditto for this change

File: geomesa-spark/geomesa_pyspark/src/main/python/setup.py [hunk 12]
```diff
@@ -0,0 +1,10 @@
+from setuptools import setup, find_packages
+
+setup(
+    name='geomesa_pyspark',
+    author='Tom Kunicki, Gerard C. Briones',
+    version='${python.version}',
+    author_email='tkunicki@ccri.com, gbriones@ccri.com',
```
Comment (jnh5y) [inline geomesa-spark/geomesa_pyspark/src/main/python/setup.py]: Is this typical?  

Is there a way to indicate geomesa.org as the Website?  And/or the geomesa-user list for questions?

File: geomesa-spark/geomesa-spark-converter/src/test/scala/org/locationtech/geomesa/spark/converter/ConverterSpatialRDDProviderTest.scala [hunk 13]
```diff
@@ -15,6 +15,7 @@ import org.geotools.data.Query
 import org.geotools.filter.text.ecql.ECQL
 import org.junit.runner.RunWith
 import org.locationtech.geomesa.spark.{GeoMesaSpark, GeoMesaSparkKryoRegistrator}
+import org.opengis.feature.simple.SimpleFeature
```
Comment (jnh5y) [inline geomesa-spark/geomesa-spark-converter/src/test/scala/org/locationtech/geomesa/spark/converter/ConverterSpatialRDDProviderTest.scala]: Is this import used?

File: geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSparkKryoRegistrator.scala [hunk 14]
```diff
@@ -87,6 +87,8 @@ object GeoMesaSparkKryoRegistrator {
 
   def register(sfts: Seq[SimpleFeatureType]): Unit = sfts.foreach(GeoMesaSparkKryoRegistrator.putTypeIfAbsent)
 
+  def register(sft: SimpleFeatureType): Unit = register(Seq(sft))
```
Comment (jnh5y) [inline geomesa-spark/geomesa-spark-core/src/main/scala/org/locationtech/geomesa/spark/GeoMesaSparkKryoRegistrator.scala]: Ok, just to be pedantic, I'd switch the implementations here;).

I think it is more natural for register(Seq) to delegate to register(sft).:)

File: geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala [hunk 15]
```diff
@@ -19,6 +19,7 @@ import org.junit.runner.RunWith
 import org.locationtech.geomesa.features.ScalaSimpleFeature
 import org.locationtech.geomesa.spark.{GeoMesaSpark, GeoMesaSparkKryoRegistrator}
 import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes
+import org.opengis.feature.simple.SimpleFeature
```
Comment (jnh5y) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: Is this import used?

File: geomesa-spark/geomesa_pyspark/pom.xml [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+            <scope>runtime</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-resources-plugin</artifactId>
+                <version>3.0.2</version>
```
Comment (jnh5y) [inline geomesa-spark/geomesa_pyspark/pom.xml]: Is the resources plugin used elsewhere?  If so, maybe we should manage the version in the root pom?

File: geomesa-spark/geomesa_pyspark/pom.xml [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+                                    <filtering>true</filtering>
+                                </resource>
+                            </resources>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+            <plugin>
+                <groupId>org.codehaus.mojo</groupId>
+                <artifactId>exec-maven-plugin</artifactId>
```
Comment (jnh5y) [inline geomesa-spark/geomesa_pyspark/pom.xml]: Are we using the exec plugin?

File: geomesa-spark/geomesa_pyspark/pom.xml [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+                        <id>regex-property</id>
+                        <phase>validate</phase>
+                        <goals>
+                            <goal>regex-property</goal>
+                        </goals>
+                        <configuration>
+                            <name>python.version</name>
+                            <value>${project.version}</value>
+                            <regex>-SNAPSHOT</regex>
+                            <replacement>.dev0</replacement>
```
Comment (jnh5y) [inline geomesa-spark/geomesa_pyspark/pom.xml]: Is 'dev0' a common Python convention?

File: geomesa-spark/geomesa_pyspark/src/main/python/.gitignore [hunk 19]
```diff
@@ -0,0 +1,3 @@
+dist/
```
Comment (jnh5y) [inline geomesa-spark/geomesa_pyspark/src/main/python/.gitignore]: Does it make sense for this to move up to the root .gitignore?

File: geomesa-spark/geomesa_pyspark/pom.xml (hunk 17 above)
Comment (tkunicki) [inline geomesa-spark/geomesa_pyspark/pom.xml]: usage in code just below this comment

File: geomesa-spark/geomesa_pyspark/src/main/python/geomesa_pyspark/__init__.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+            raise ImportError('unable to locate ' + executor_package + ' installed in driver')
+
+        package = sys.modules.get(executor_package, None)
+        if package is None:
+            package = pkgutil.get_loader(executor_package).load_module(executor_package)
+
+        package_path = os.path.dirname(package.__file__)
+        package_root = os.path.dirname(package_path)
+
+        if package_root[-4:].lower() in PACKAGE_EXTENSIONS:
```
Comment (jnh5y) [inline geomesa-spark/geomesa_pyspark/src/main/python/geomesa_pyspark/__init__.py]: Can you add a comment for the '-4'? 

I *think* this is doing something like cutting off the last few things in the classpath...  Something about this seems fragile...

File: geomesa-spark/geomesa_pyspark/pom.xml (hunk 18 above)
Comment (tkunicki) [inline geomesa-spark/geomesa_pyspark/pom.xml]: yes, [PEP 440 -- Version Identification and Dependency Specification](https://www.python.org/dev/peps/pep-0440/)

File: geomesa-spark/geomesa_pyspark/src/main/python/geomesa_pyspark/__init__.py (hunk 20 above)
Comment (tkunicki) [inline geomesa-spark/geomesa_pyspark/src/main/python/geomesa_pyspark/__init__.py]: this is 'pythonic', grabbed from pyspark source

File: geomesa-spark/geomesa_pyspark/src/main/python/.gitignore (hunk 19 above)
Comment (tkunicki) [inline geomesa-spark/geomesa_pyspark/src/main/python/.gitignore]: some of the stuff is localized being our first python module, can be moved if we want to embrace...

File: geomesa-spark/geomesa_pyspark/pom.xml (hunk 16 above)
Comment (tkunicki) [inline geomesa-spark/geomesa_pyspark/pom.xml]: yep, can move up...  wasn't sure how were were going to handle maven module parenting w/ python

File: geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala (hunk 15 above)
Comment (tkunicki) [inline geomesa-spark/geomesa-spark-geotools/src/test/scala/org/locationtech/geomesa/spark/geotools/GeoToolsSpatialRDDProviderTest.scala]: looks like it's unused w/ refactor, can remove

File: geomesa-hbase/geomesa-hbase-tools/src/main/resources/org/locationtech/geomesa/hbase/tools/ingest-libjars.list (hunk 10 above)
Comment (tkunicki) [inline geomesa-hbase/geomesa-hbase-tools/src/main/resources/org/locationtech/geomesa/hbase/tools/ingest-libjars.list]: i didn't commit

File: geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/IndexAdapter.scala (hunk 11 above)
Comment (tkunicki) [inline geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/IndexAdapter.scala]: i didn't commit

File: geomesa-spark/geomesa_pyspark/src/main/python/setup.py (hunk 12 above)
Comment (tkunicki) [inline geomesa-spark/geomesa_pyspark/src/main/python/setup.py]: we can review how we want to handle this.

File: geomesa-spark/geomesa-spark-converter/src/test/scala/org/locationtech/geomesa/spark/converter/ConverterSpatialRDDProviderTest.scala (hunk 13 above)
Comment (tkunicki) [inline geomesa-spark/geomesa-spark-converter/src/test/scala/org/locationtech/geomesa/spark/converter/ConverterSpatialRDDProviderTest.scala]: looks dead to me, can remove
